pretyping/inferCumulativity.ml

```diff
@@ -0,0 +1,245 @@
+   restrict it accordingly. *)
+
+let infer_level_eq u variances =
+  if LMap.mem u variances
+  then LMap.set u Variance.Invariant variances
+  else variances
+
+let infer_level_leq u variances =
+  match LMap.find u variances with
+  | exception Not_found -> variances
+  | varu -> LMap.set u (Variance.sup varu Variance.Covariant) variances
+
+let infer_generic_instance_eq variances u =
+  Array.fold_left (fun variances u -> infer_level_eq u variances)
+    variances (Instance.to_array u)
+
+let variance_pb cv_pb var =
+  let open Variance in
+  match cv_pb, var with
+  | _, Irrelevant -> Irrelevant
+  | _, Invariant -> Invariant
+  | CONV, Covariant -> Invariant
+  | CUMUL, Covariant -> Covariant
+
+let infer_cumulative_ind_instance cv_pb cumi variances u =
+  Array.fold_left2 (fun variances varu u ->
+      match LMap.find u variances with
+      | exception Not_found -> variances
+      | varu' ->
+        LMap.set u (Variance.sup varu' (variance_pb cv_pb varu)) variances)
+    variances (ACumulativityInfo.variance cumi) (Instance.to_array u)
+
+let infer_inductive_instance cv_pb env variances ind nargs u =
+  let mind = Environ.lookup_mind (fst ind) env in
+  match mind.mind_universes with
+  | Monomorphic_ind _ -> assert (Instance.is_empty u); variances
+  | Polymorphic_ind _ -> infer_generic_instance_eq variances u
+  | Cumulative_ind cumi ->
+    if not (Int.equal (inductive_cumulativity_arguments (mind,snd ind)) nargs)
+    then infer_generic_instance_eq variances u
+    else infer_cumulative_ind_instance cv_pb cumi variances u
+
+let infer_constructor_instance_eq env variances ((mi,ind),ctor) nargs u =
+  let mind = Environ.lookup_mind mi env in
+  match mind.mind_universes with
+  | Monomorphic_ind _ -> assert (Instance.is_empty u); variances
+  | Polymorphic_ind _ -> infer_generic_instance_eq variances u
+  | Cumulative_ind cumi ->
+    if not (Int.equal (constructor_cumulativity_arguments (mind,ind,ctor)) nargs)
+    then infer_generic_instance_eq variances u
+    else infer_cumulative_ind_instance CONV cumi variances u
+
+let infer_sort cv_pb variances s =
+  match cv_pb with
+  | CONV ->
+    LSet.fold infer_level_eq (Universe.levels (Sorts.univ_of_sort s)) variances
+  | CUMUL ->
+    LSet.fold infer_level_leq (Universe.levels (Sorts.univ_of_sort s)) variances
+
+let is_invariant_instance variances u =
+  let open Variance in
+  Instance.is_empty u || Array.for_all (fun u ->
+      match LMap.find u variances with
+      | Invariant -> true
+      | exception Not_found -> true
+      | Covariant | Irrelevant -> false)
+    (Instance.to_array u)
+
+type foo =
+  | NoUnfolding of Variance.t LMap.t
+  | Unfolding of CClosure.fconstr
+
+let infer_table_key infos variances =
+  let open Names in
+  function
+  | ConstKey (cst, u) as c ->
+    let flex = Environ.evaluable_constant cst (CClosure.info_env infos) in
+    if flex
+    then
+      if is_invariant_instance variances u
```

Comment:
After thinking about it I don't trust this heuristic to be good enough. eg if we get `id Type nat` (with monomorphic `id`) the universe on the `Type` will be thought irrelevant when it shouldn't be.
We could reproduce what reduction.ml does by trying to infer_stack without unfolding and if we encounter a local universe backtrack and unfold, but if we always reduce 3330.v is still only at 1s (vs 0.8 with the heuristic) so I propose we always reduce constants.

Reply:
You mean the heuristic with the ULub’s that tries to avoid unfolding if only flexible levels are involved?

If not needed I agree it’s simpler and safer to not use a heuristic

Reply:
No, this this bit: https://github.com/coq/coq/blob/52d666a7a83e4023d9f5cd7324ed81c7f7926156/kernel/reduction.ml#L409-L412